### DISCUSSION TITLE
Performance of kong with http-log

### DISCUSSION THREAD

**Discussion by Sachin-M-Desai** [Kong/kong#10748] [Help]
Hi,

We are having to run performance bechmarking on kong:3.2.2 (doesnt matter) with http-log. Here are some numbers with wrk tool. The setup isnt very highly loaded with 2 dedicated CPUs provided to kong and KONG_NGINX_WORKER_PROCESSES=4.

With no http-log enabled,
Base performance without plugin
--------------------------------
ubuntu@ip-10-101-1-31:~$ wrk -t 1 -d 10s -c 2 --latency  -s post-sess-large.lua http://http-service:9500/setprofile  
Running 10s test @ http://http-service:9500/setprofile                                                               
  1 threads and 2 connections                                                                                        
  Thread Stats   Avg      Stdev     Max   +/- Stdev                                                                  
    Latency     0.91ms  443.46us  13.90ms   94.12%                                                                   
    Req/Sec     2.24k   211.74     2.52k    81.00%                                                                   
  Latency Distribution                                                                                               
     50%  835.00us                                                                                                   
     75%    1.00ms                                                                                                   
     90%    1.21ms                                                                                                   
     99%    2.23ms                                                                                                   
  22281 requests in 10.00s, 5.14MB read                                                                              
Requests/sec:   2227.85                                                                                              
Transfer/sec:    526.49KB       

http-log performance
--------------------
ubuntu@ip-10-101-1-31:~$ wrk -t 1 -d 10s -c 2 --latency  -s post-sess-large.lua http://http-service:9500/setprofile       
Running 10s test @ http://http-service:9500/setprofile                                                                    
  1 threads and 2 connections                                                                                             
  Thread Stats   Avg      Stdev     Max   +/- Stdev                                                                       
    Latency     1.48ms  668.65us  17.28ms   90.98%                                                                        
    Req/Sec     1.39k   208.13     2.14k    85.00%                                                                        
  Latency Distribution                                                                                                    
     50%    1.39ms                                                                                                        
     75%    1.58ms                                                                                                        
     90%    1.85ms                                                                                                        
     99%    3.73ms                                                                                                        
  13817 requests in 10.01s, 3.19MB read                                                                                   
Requests/sec:   1380.98                                                                                                   
Transfer/sec:    326.36KB   

While the test is running the nginx workers arent heavily loaded either (CPU max ~85%)

This is a large drop in the performance considering the logging to remote endpoint happens asynchronously. 

This has been our findings. The bottleneck is the resty.http library which don't scale well and bring down the performance (adds to latency) even though, executed in the ngx.timer context.

So questions,
1. Are there any better alternatives to sending logs asynchronously?
2. Any known solutions to this?

**Comment by samugi** [comment]
Hi @Sachin-M-Desai 
thank you for opening this discussion.

> While the test is running the nginx workers arent heavily loaded either

What about memory utilization? My thinking is that the large amount of timers created might be affecting performance.

In general, adding a dependency to an external service (although reached asynchronously) makes it difficult to compare performance meaningfully. You might want to repeat the test using a mock (local) log server and see how that affects performance and resources utilization. 

That said, this plugin relies heavily on batch queues, which have [recently been reimplemented](https://github.com/Kong/kong/pull/10172), so my suggestion would be to test this behaviour again on 3.3 once released, or on the current master branch if you'd like to get a preview of how the new implementation affects performance.

**Comment by Sachin-M-Desai** [comment]
> What about memory utilization? My thinking is that the large amount of timers created might be affecting performance.

The memory utilization wasnt much. I had allocated 2G to the kong docker, but the usage was less than 500MB. 

> In general, adding a dependency to an external service (although reached asynchronously) makes it difficult to compare performance meaningfully.

The http_endpoint and backend server are all on the same local network and CPU isolated (via docker though). The http_endpoint is a dummy one with no latency (just replies with 200 OK).

I ended up repeating the same experiement by adding more CPUs to kong, but there isnt much of an improvement. 

Is it possible to perform the resty.http calls (httpc.connect and httpc.request) via threads? Timers aren't scalable and the performance comes down heavily if nginx workers hit high CPU (99%).

My concern here is that, while we say the logging is asynchronous, the overall latency and throughput is badly affected. 

What other alternatives exist today? Can there be a truly threaded implementation we can use to send asynchronously?

**Reply by samugi** [reply]
my recommendation would be to test the reimplemented queues as mentioned in my previous comment, and the tuning that is provided via the newly added configuration parameters such as `max_batch_size`. This change alone is likely to affect performance heavily so I believe it is important to re-evaluate with the new implementation first.

**Reply by Sachin-M-Desai** [reply]
Thanks. If possibly would be possible to share any initial performance results?

**Reply by hanshuebner** [reply]
@Sachin-M-Desai In our tests, the performance of the http-log plugin with the new queue implementation was generally better than with the old implementation.  I'd recommend that you repeat your test with Kong 3.3 (or the current master branch) to see if it helps in your situation.  The `queue.max_batch_size` parameter may be worth setting to a higher value to enable batching of multiple log requests.
